erts/emulator/beam/jit/x86/instr_arith.cpp

```diff
@@ -841,115 +900,263 @@ void BeamGlobalAssembler::emit_times_guard_shared() {
+        emit_leave_frame();
+
+        emit_test_the_non_value(RET);
+        a.short_().je(mul_error);
+
+        a.ret();
+    }
+
     a.bind(error);
     {
-        /* Place the original arguments in x-registers. */
+        static const ErtsCodeMFA mul_mfa = {am_erlang, am_Times, 2};
+        static const ErtsCodeMFA add_mfa = {am_erlang, am_Plus, 2};
+
+        a.mov(ARG1, TMP_MEM3q);
+        a.mov(ARG2, TMP_MEM4q);
+        mov_imm(ARG4, &add_mfa);
+        emit_test_the_non_value(ARG1);
+        a.short_().jne(do_error);
+
+        a.bind(mul_error);
         a.mov(ARG1, TMP_MEM1q);
         a.mov(ARG2, TMP_MEM2q);
+        mov_imm(ARG4, &mul_mfa);
+
+        a.bind(do_error);
         a.mov(getXRef(0), ARG1);
         a.mov(getXRef(1), ARG2);
-
-        a.mov(ARG4, imm(&bif_mfa));
         a.jmp(labels[raise_exception]);
     }
 }
 
-void BeamModuleAssembler::emit_i_times(const ArgLabel &Fail,
-                                       const ArgSource &LHS,
-                                       const ArgSource &RHS,
-                                       const ArgRegister &Dst) {
-    bool small_result = is_product_small_if_args_are_small(LHS, RHS);
+/* ARG2 = Src1
+ * ARG3 = Src2
+ *
+ * The result is returned in RET.
+ */
+void BeamGlobalAssembler::emit_mul_body_shared() {
+    mov_imm(ARG4, make_small(0));
+    a.jmp(labels[mul_add_body_shared]);
+}
 
-    if (always_small(LHS) && always_small(RHS) && small_result) {
-        comment("multiplication without overflow check");
-        if (RHS.isSmall()) {
-            Sint factor = RHS.as<ArgSmall>().getSigned();
+/* ARG2 = Src1
+ * ARG3 = Src2
+ *
+ * Result is returned in RET, error is indicated by ZF.
+ */
+void BeamGlobalAssembler::emit_mul_guard_shared() {
+    mov_imm(ARG4, make_small(0));
+    a.jmp(labels[mul_add_guard_shared]);
+}
+
+void BeamModuleAssembler::emit_i_mul_add(const ArgLabel &Fail,
+                                         const ArgSource &Src1,
+                                         const ArgSource &Src2,
+                                         const ArgSource &Src3,
+                                         const ArgSource &Src4,
+                                         const ArgRegister &Dst) {
+    bool is_product_small = is_product_small_if_args_are_small(Src1, Src2);
+    bool is_sum_small = is_sum_small_if_args_are_small(Src3, Src4);
+    bool is_increment_zero =
+            Src4.isSmall() && Src4.as<ArgSmall>().getSigned() == 0;
+    Sint factor = 0;
+    int left_shift = -1;
+
+    if (is_increment_zero) {
+        comment("(adding zero)");
+    }
+
+    if (Src2.isSmall()) {
+        factor = Src2.as<ArgSmall>().getSigned();
+        if (Support::isPowerOf2(factor)) {
+            left_shift = Support::ctz<Eterm>(factor);
+        }
+    }
+
+    if (always_small(Src1) && Src2.isSmall() && Src4.isSmall() &&
+        is_product_small && is_sum_small) {
+        x86::Mem p;
+        Sint increment = Src4.as<ArgSmall>().get();
```

Comment:
If I'm not mistaken, we can skip untagging by letting `increment` cancel out the multiplied tag bits.

Reply:
Yes, that should work. I'll try to do it on Monday.

Reply:
I've pushed a commit that skips untagging.